[Flaky Test] `PodZoneAffinity` Integration Test Suite

<!-- Please only use this template for submitting reports about flaky tests or jobs (pass or fail with no underlying change in code) in Gardener CI -->

**How to categorize this issue?**
<!--
Please select area, kind, and priority for this issue. This helps the community categorizing it.
Replace below TODOs or exchange the existing identifiers with those that fit best in your opinion.
If multiple identifiers make sense you can also state the commands multiple times, e.g.
  /area control-plane
  /area auto-scaling
  ...

"/area" identifiers:     audit-logging|auto-scaling|backup|certification|control-plane-migration|control-plane|cost|delivery|dev-productivity|disaster-recovery|documentation|high-availability|logging|metering|monitoring|networking|open-source|ops-productivity|os|performance|quality|robustness|scalability|security|storage|testing|usability|user-management
"/kind" identifiers:     api-change|bug|cleanup|discussion|enhancement|epic|impediment|poc|post-mortem|question|regression|task|technical-debt|test
-->
/area testing
/kind flake

**Which test(s)/suite(s) are flaking**:
`PodZoneAffinity` Integration Test Suite

**CI link**:
- https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6647/pull-gardener-integration/1567724664074014720

**Reason for failure**:
```
• [FAILED] [0.014 seconds]
PodSchedulerName tests
/home/prow/go/src/github.com/gardener/gardener/test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go:26
  when namespace has zone enforcement label with value
  /home/prow/go/src/github.com/gardener/gardener/test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go:81
    [It] should add nodeAffinity
    /home/prow/go/src/github.com/gardener/gardener/test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go:96
  Begin Captured GinkgoWriter Output >>
    {"level":"debug","ts":"2022-09-08T04:04:05.188Z","logger":"controller-runtime.webhook.webhooks","msg":"received request","webhook":"/webhooks/pod-zone-affinity","UID":"b2cd2da0-b6c1-4c5f-aa06-baf99d7c5c60","kind":"/v1, Kind=Pod","resource":{"group":"","version":"v1","resource":"pods"}}
    {"level":"debug","ts":"2022-09-08T04:04:05.188Z","logger":"controller-runtime.webhook.webhooks","msg":"wrote response","webhook":"/webhooks/pod-zone-affinity","code":200,"reason":"","UID":"b2cd2da0-b6c1-4c5f-aa06-baf99d7c5c60","allowed":true}
  << End Captured GinkgoWriter Output
  Expected
      <*v1.NodeAffinity | 0x0>: nil
  not to be nil
  In [It] at: /home/prow/go/src/github.com/gardener/gardener/test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go:101
------------------------------
•
Ran 3 of 3 Specs in 6.686 seconds
FAIL! -- 2 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestPodPodZoneAffinity (6.69s)
FAIL
FAIL	github.com/gardener/gardener/test/integration/resourcemanager/podzoneaffinity	6.799s
```

**Anything else we need to know**:
Introduced with #6579 and first seen in https://github.com/gardener/gardener/pull/6647#issuecomment-1240193626

--- a/test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_suite_test.go
+++ b/test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_suite_test.go
@@ -69,7 +69,7 @@ var _ = BeforeSuite(func() {
 	By("starting test environment")
 	testEnv = &envtest.Environment{
 		WebhookInstallOptions: envtest.WebhookInstallOptions{
-			MutatingWebhooks: getMutatingWebhookConfigurations(testNamespaceName),
+			MutatingWebhooks: getMutatingWebhookConfigurations(),
 		},
 	}
 
@@ -109,6 +109,10 @@ var _ = BeforeSuite(func() {
 		CertDir:            testEnv.WebhookInstallOptions.LocalServingCertDir,
 		MetricsBindAddress: "0",
 		Namespace:          testNamespace.Name,
+		ClientDisableCacheFor: []client.Object{
+			// Disable cache for namespaces so that changes applied by tests are seen immediately.
+			&corev1.Namespace{},
+		},
 	})
 	Expect(err).NotTo(HaveOccurred())
 
@@ -135,7 +139,7 @@ var _ = BeforeSuite(func() {
 	})
 })
 
-func getMutatingWebhookConfigurations(namespace string) []*admissionregistrationv1.MutatingWebhookConfiguration {
+func getMutatingWebhookConfigurations() []*admissionregistrationv1.MutatingWebhookConfiguration {
 	webhookConfig := []*admissionregistrationv1.MutatingWebhookConfiguration{
 		{
 			TypeMeta: metav1.TypeMeta{
@@ -156,12 +160,6 @@ func getMutatingWebhookConfigurations(namespace string) []*admissionregistration
 			},
 		},
 	}
-
-	webhookConfig[0].Webhooks[0].NamespaceSelector.MatchExpressions = append(webhookConfig[0].Webhooks[0].NamespaceSelector.MatchExpressions, metav1.LabelSelectorRequirement{
-		Key:      corev1.LabelMetadataName,
-		Operator: metav1.LabelSelectorOpIn,
-		Values:   []string{namespace},
-	})
 
 	return webhookConfig
 }

--- a/test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go
+++ b/test/integration/resourcemanager/podzoneaffinity/podzoneaffinity_test.go
@@ -23,7 +23,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-var _ = Describe("PodSchedulerName tests", func() {
+var _ = Describe("PodZoneAffinity tests", func() {
 	var pod *corev1.Pod
 
 	BeforeEach(func() {
@@ -50,14 +50,12 @@ var _ = Describe("PodSchedulerName tests", func() {
 	Context("when namespace has zone enforcement label", func() {
 		BeforeEach(func() {
 			patch := client.MergeFrom(testNamespace.DeepCopy())
-			testNamespace.Labels = map[string]string{
-				"control-plane.shoot.gardener.cloud/enforce-zone": "",
-			}
+			testNamespace.Labels["control-plane.shoot.gardener.cloud/enforce-zone"] = ""
 			Expect(testClient.Patch(ctx, testNamespace, patch)).To(Succeed())
 
 			DeferCleanup(func() {
 				patch := client.MergeFrom(testNamespace.DeepCopy())
-				testNamespace.Labels = nil
+				delete(testNamespace.Labels, "control-plane.shoot.gardener.cloud/enforce-zone")
 				Expect(testClient.Patch(ctx, testNamespace, patch)).To(Succeed())
 			})
 		})
@@ -81,14 +79,12 @@ var _ = Describe("PodSchedulerName tests", func() {
 	Context("when namespace has zone enforcement label with value", func() {
 		BeforeEach(func() {
 			patch := client.MergeFrom(testNamespace.DeepCopy())
-			testNamespace.Labels = map[string]string{
-				"control-plane.shoot.gardener.cloud/enforce-zone": "zone-a",
-			}
+			testNamespace.Labels["control-plane.shoot.gardener.cloud/enforce-zone"] = "zone-a"
 			Expect(testClient.Patch(ctx, testNamespace, patch)).To(Succeed())
 
 			DeferCleanup(func() {
 				patch := client.MergeFrom(testNamespace.DeepCopy())
-				testNamespace.Labels = nil
+				delete(testNamespace.Labels, "control-plane.shoot.gardener.cloud/enforce-zone")
 				Expect(testClient.Patch(ctx, testNamespace, patch)).To(Succeed())
 			})
 		})

